A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,242 @@
+A48: xDS Least Request LB Policy
+----
+* Author(s): erikjoh & [fabric](https://github.com/orgs/spotify/teams/fabric)
+* Approver: markdroth
+* Status: Implementation in progress
+* Implemented in: Java in progress
+* Last updated: 2021-11-12
+* Discussion at: https://groups.google.com/g/grpc-io/c/4qycdcFfMUs
+
+## Abstract
+
+Add support for least request load balancing configured via xDS.
+
+## Background
+
+In multi-tenancy environments like for example Kubernetes, some upstream endpoints may perform much slower than others.
+When using plain round-robin load balancing, this becomes a problem since all endpoints get an equal amount of RPCs
+even when some endpoints perform worse. The result is increased tail latencies and in extreme cases even increased error rates.
+As an alternative to round-robin, Envoy provides a "least request"
+[load balancing policy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request)
+which uses per-endpoint outstanding/active request counters as a heuristic for determining load on each endpoint.
+The Envoy least request implementation uses two different algorithms depending on whether endpoints have non-equal
+weights or not. In the "*all weights equal*" case, Envoy samples N random available hosts and selects the one with
+fewest outstanding requests. In the "*all weights not equal*" case, Envoy switches to a weighted round robin schedule
+where weights are dynamically adjusted based on the outstanding requests.
+
+### Related Proposals: 
+
+This proposal builds on earlier work described in the following gRFCs:
+* [gRFC A27: xDS-based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+* [gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md)
+
+## Proposal
+
+Support `LEAST_REQUEST` LB policy and configuration of it using xDS.
+Specifically, gRPC will only support the "*all weights equal*" least request algorithm as
+[defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request).
+
+### Least Request LB Policy
+
+#### xDS API Fields
```

Comment:
This section does not mention the `slow_start_config` field.  Is that going to be supported?

Reply:
Good catch! Was looking at an old version of the proto which didn't have that field.
This feature requires the Weighted Least Request algorithm so if we don't support that initially we wouldn't support `slow_start_config` either.
Also, I think support for this might be a bit tricky for the "*all weights equal*" case since Envoy seems to switch to the "*all weights not equal*" algorithm during endpoint startup and then back to the "*all weights equal*" algorithm once all endpoints have fully started. This might get a bit complicated in gRPC as it means the picker behavior would need to change in between resolves.

Reply:
Sounds good.  But let's explicitly state that it won't be supported in the doc, so that there's no confusion.